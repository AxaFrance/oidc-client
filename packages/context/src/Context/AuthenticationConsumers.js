import React, { Fragment } from 'react';
import { withRouter } from 'react-router-dom';
import {
  fromRenderProps,
  compose,
  branch,
  lifecycle,
  renderComponent,
} from 'recompose';
import {
  AuthenticationConsumer,
  withOidcUser,
} from './AuthenticationContext.container';
import {
  authenticateUser,
  getUserManager,
  oidcLog,
  isRequireAuthentication,
} from '../Services';
import { Authenticating } from '../OidcComponents';

const withContext = fromRenderProps(
  AuthenticationConsumer,
  ({ isEnabled }) => ({ isEnabled }),
);

const lifecycleComponent = {
  async componentDidMount() {
    oidcLog.info('Protected component mounted');
    const usermanager = getUserManager();
    await authenticateUser(usermanager, this.props.location)();
  },
};

const wrapAuthenticating = () => <Authenticating />;
const Dummy = ({ children }) => <Fragment>{children}</Fragment>;

export const withOidcSecure = compose(
  withContext,
  branch(({ isEnabled }) => !isEnabled, renderComponent(Dummy)),
  withOidcUser,
  withRouter,
  lifecycle(lifecycleComponent),
  branch(
    ({ oidcUser }) => isRequireAuthentication(oidcUser, false),
    renderComponent(wrapAuthenticating),
  ),
);

const OidcSecure = props => {
  const { children } = props;
  const Secure = withOidcSecure(Dummy);
  return <Secure {...props}>{children}</Secure>;
};

export default OidcSecure;
