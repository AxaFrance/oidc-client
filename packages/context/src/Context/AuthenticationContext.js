import React from 'react';
import PropTypes from 'prop-types';
import { AuthenticationContext } from './AuthenticationContextCreator';
import { OidcRoutes } from '../Routes';

const propTypes = {
  notAuthentified: PropTypes.node,
  notAuthorized: PropTypes.node,
  isLoading: PropTypes.bool.isRequired,
  isEnabled: PropTypes.bool,
  // eslint-disable-next-line
  oidcUser: PropTypes.object, //TODO : rÃ©cuperer le proptypes depuis OIDC client ?
  error: PropTypes.string.isRequired,
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

const defaultProps = {
  notAuthentified: null,
  notAuthorized: null,
  isEnabled: true,
};

const AuthenticationProviderComponent = ({
  isLoading,
  isEnabled,
  oidcUser,
  error,
  login,
  logout,
  notAuthentified,
  notAuthorized,
  children,
}) => (
  <AuthenticationContext.Provider
    value={{
      isLoading,
      oidcUser,
      error,
      login,
      logout,
      isEnabled,
    }}
  >
    <OidcRoutes notAuthentified={notAuthentified} notAuthorized={notAuthorized}>
      {children}
    </OidcRoutes>
  </AuthenticationContext.Provider>
);

AuthenticationProviderComponent.propTypes = propTypes;
AuthenticationProviderComponent.defaultProps = defaultProps;

export default AuthenticationProviderComponent;
