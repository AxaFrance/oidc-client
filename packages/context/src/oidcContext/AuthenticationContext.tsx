import React, { ComponentType } from 'react';
import { User, UserManagerEvents, UserManager } from 'oidc-client';

export type oidcContext = {
  oidcUser: User | null;
  isEnabled: boolean;
  login: Function;
  loginSilent: Function;
  loginPopup: Function;
  logout: Function;
  events: UserManagerEvents;
  authenticating: ComponentType;
  isLoading: boolean;
  isLoggingOut: boolean;
  userManager: UserManager;
  error: string;
};

export const AuthenticationContext = React.createContext<oidcContext>(null);

export const useReactOidc = () => {
  const { isEnabled, login, loginSilent, loginPopup, logout, oidcUser, events } = React.useContext(AuthenticationContext);
  return { isEnabled, login, loginSilent, loginPopup, logout, oidcUser, events };
};
