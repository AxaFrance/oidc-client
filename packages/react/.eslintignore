build
bin
node_modules
dist
**/*.spec.ts
**/*.spec.tsx
**/*.test.ts
**/*.test.tsx
**/*.stories.tsx
./src/setupTests.ts
setupTests.js
.eslintrc.js
config
jest.config.js
App.specold.tsx
