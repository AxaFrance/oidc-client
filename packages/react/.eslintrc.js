module.exports = {
    extends: ['./config/defaultEslintConfig'], 
    parserOptions: {
      project: './tsconfig.eslint.json',
      tsconfigRootDir: __dirname,
    },
    rules: {
      '@typescript-eslint/naming-convention': [
        'error',
        {
          'selector': 'variable',
          'types': ['boolean'],
          'format': ['PascalCase'],
          'prefix': ['is', 'with', 'should', 'has', 'can', 'did', 'will']
        }
      ]
    }
  }