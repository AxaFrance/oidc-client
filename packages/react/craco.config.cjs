module.exports = {
    webpack: {
      configure: (webpackConfig, { env, paths }) => {
        webpackConfig.resolve.extensionAlias = {
          ".js": [".ts", ".js", ".tsx", ".jsx"],
          ".mjs": [".mts", ".mjs"]
        }
        return webpackConfig;
      },
    },
  };