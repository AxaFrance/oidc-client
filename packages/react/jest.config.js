module.exports = {
    collectCoverage: true,
    collectCoverageFrom: ['src/**/*.{ts,tsx}'],
};