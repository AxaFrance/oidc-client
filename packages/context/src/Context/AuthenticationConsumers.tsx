import React, { ComponentType, PropsWithChildren, useContext, useEffect, useMemo } from 'react';
import {
  withRouter,
  Authenticating,
  getUserManager,
  authenticateUser,
  isRequireAuthentication,
  oidcLog,
  ReactOidcHistory,
} from '@axa-fr/react-oidc-core';
import withServices from '../withServices';
import { AuthenticationContext } from './AuthenticationContext';

// export use only for unit tests
export const useOidcSecure = (
  authenticateUserInternal: typeof authenticateUser,
  getUserManagerInternal: typeof getUserManager,
  location: Location,
  history: ReactOidcHistory
) => {
  const { isEnabled, oidcUser, authenticating } = useContext(AuthenticationContext);
  useEffect(() => {
    oidcLog.info('Protection : ', isEnabled);
    if (isEnabled) {
      oidcLog.info('Protected component mounted');
      const usermanager = getUserManagerInternal();
      authenticateUserInternal(usermanager, location, history)();
    }
    return () => {
      oidcLog.info('Protected component unmounted');
    };
  }, [location, isEnabled, authenticateUserInternal, getUserManagerInternal]);
  return { oidcUser, authenticating, isEnabled };
};

// for usage
export const useReactOidc = () => {
  const { isEnabled, login, logout, oidcUser, events } = useContext(AuthenticationContext);
  return { isEnabled, login, logout, oidcUser, events };
};

type WithRouterComponentProps = PropsWithChildren<{
  location: Location;
  history: ReactOidcHistory;
}>;

const OidcSecure = withRouter(({ children, location, history }: WithRouterComponentProps) => {
  const { oidcUser, authenticating, isEnabled } = useOidcSecure(authenticateUser, getUserManager, location, history);
  const requiredAuth = useMemo(() => isRequireAuthentication(oidcUser, false) && isEnabled, [isEnabled, oidcUser]);
  const AuthenticatingComponent = authenticating || Authenticating;
  return requiredAuth ? <AuthenticatingComponent /> : <div>{children}</div>;
});

export default OidcSecure;

// For non-regression
type WithOidcSecurewithRouterProps = PropsWithChildren<{
  location: Location;
  history: ReactOidcHistory;
  authenticateUser: typeof authenticateUser;
  getUserManager: typeof getUserManager;
}>;

export const withOidcSecurewithRouter = (WrappedComponent: ComponentType) => ({
  location,
  history,
  authenticateUser: authenticateUserInternal,
  getUserManager: getUserManagerInternal,
  ...otherProps
}: WithOidcSecurewithRouterProps) => {
  const { oidcUser, authenticating, isEnabled } = useOidcSecure(authenticateUserInternal, getUserManagerInternal, location, history);
  const requiredAuth = useMemo(() => isRequireAuthentication(oidcUser, false) && isEnabled, [isEnabled, oidcUser]);

  const AuthenticatingComponent = authenticating || Authenticating;
  return requiredAuth ? <AuthenticatingComponent /> : <WrappedComponent {...otherProps} />;
};

export const withOidcSecure = (WrappedComponent: ComponentType) =>
  withRouter(
    withServices(withOidcSecurewithRouter(WrappedComponent), {
      authenticateUser,
      getUserManager,
    })
  );

type WithOidcUserComponentProps = PropsWithChildren<{}>;
export const withOidcUser = (Component: ComponentType) => (props: WithOidcUserComponentProps) => {
  const { oidcUser } = useContext(AuthenticationContext);
  const { children } = props;
  return (
    // @ts-ignore
    <Component {...props} oidcUser={oidcUser}>
      {children}
    </Component>
  );
};
