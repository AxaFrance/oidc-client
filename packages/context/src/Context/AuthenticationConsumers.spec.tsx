import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useOidcSecure, withOidcUser, withOidcSecurewithRouter } from './AuthenticationConsumers';
import { AuthenticationContext } from './AuthenticationContext';

describe('Consumers service tests suite', () => {
  const values = {
    isEnabled: true,
    oidcUser: { user: 'tom' },
    authenticating: { param: 'test' },
  };

  const getWrapper = (vals = values) => ({ children }) => (
    <AuthenticationContext.Provider value={vals}>{children}</AuthenticationContext.Provider>
  );
  const getUserManager = () => ({
    param: 'Mock version',
  });
  const authenticateUser = jest.fn(() => jest.fn());

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should call authentication when enabled', () => {
    const history = { push: () => {} };
    renderHook(() => useOidcSecure(authenticateUser, getUserManager, '/locationUser', history), {
      wrapper: getWrapper(),
    });

    expect(authenticateUser).toHaveBeenCalledWith({ param: 'Mock version' }, '/locationUser', history);
  });

  it('should NOT call authentication when disabled', () => {
    renderHook(() => useOidcSecure(authenticateUser, getUserManager, '/locationUser'), {
      wrapper: getWrapper({ ...values, isEnabled: false }),
    });
    expect(authenticateUser).not.toHaveBeenCalled();
  });

  it('withOidcUser should inject oidcUser if user connected', () => {
    const testingComponent = ({ oidcUser }) => (
      <div>
        <span data-testid="username">{oidcUser}</span>
      </div>
    );
    const Component = withOidcUser(testingComponent);
    const { getByTestId, asFragment } = render(<Component />, {
      wrapper: getWrapper({ oidcUser: 'Franck' }),
    });
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId('username')).toHaveTextContent('Franck');
  });

  it('withOidcSecurewithRouter should call authentication and AuthenticatingComponent during authentication process', () => {
    const contextMock = {
      isEnabled: true,
      oidcUser: null,
    };

    const mockedProps = {
      location: '/myurl',
      authenticateUser,
      getUserManager,
      history: { push: () => {} },
    };

    const testingComponent = () => <div data-testid="componentMount" />;
    const Component = withOidcSecurewithRouter(testingComponent);
    const { queryByTestId, asFragment } = render(<Component {...mockedProps} />, {
      wrapper: getWrapper({ ...contextMock }),
    });
    expect(asFragment()).toMatchSnapshot();
    expect(queryByTestId('componentMount')).toBeNull();
  });

  it('withOidcSecurewithRouter should show component if user connected', () => {
    const contextMock = {
      isEnabled: true,
      oidcUser: { user: 'tom' },
    };

    const mockedProps = {
      location: '/myurl',
      authenticateUser,
      getUserManager,
      history: { push: () => {} },
    };

    const testingComponent = () => <div data-testid="componentMount" />;
    const Component = withOidcSecurewithRouter(testingComponent);
    const { queryByTestId } = render(<Component {...mockedProps} />, {
      wrapper: getWrapper({ ...contextMock }),
    });
    expect(queryByTestId('componentMount')).not.toBeNull();
  });
});
