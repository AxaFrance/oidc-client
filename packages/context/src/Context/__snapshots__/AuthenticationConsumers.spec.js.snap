// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consumers service tests suite withOidcSecurewithRouter should call authentication and AuthenticatingComponent during authentication process 1`] = `
<DocumentFragment>
  <div
    class="oidc-authenticating"
  >
    <div
      class="oidc-authenticating__container"
    >
      <h1
        class="oidc-authenticating__title"
      >
        Authentification en cours
      </h1>
      <p
        class="oidc-authenticating__content"
      >
        Vous allez être redirigé sur la page de login
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Consumers service tests suite withOidcUser should inject oidcUser if user connected 1`] = `
<DocumentFragment>
  <div>
    <span
      data-testid="username"
    >
      Franck
    </span>
  </div>
</DocumentFragment>
`;
